caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -96,6 +96,12 @@
 
     private PathGenerator pathGenerator;
 
+    // This thread holder is used to add some "re-entrant" like function for the ISPN transaction lock. The ISPN lock is
+    // used for ISPN transaction but not for thread level with re-entrant, that means if we want to own this lock in one
+    // transaction for more than twice in single thread, it will block this thread. So we need this thread holder to by-pass
+    // the ISPN lock waiting when thread not changed.
+    private final ThreadLocal<Long> threadHolder = new ThreadLocal<>();
```

Comment:
I have two concerns here:

1. What will clear this ThreadLocal? If you're acting only through the CacheProvider interface, and responding to a close() call to clear this, how do you know whether you're dealing with the close() call for a derivative (eg. checksum) file, or the original file that caused the lock to be created?

2. Can a single thread have two open files in different directories? If so, will one overwrite the ThreadLocal, and make it even harder to know when to clear it? 

Usually in cases like these, we resort to using reference counting, so that each new lock/open operation increases the lock count, and each subsequent unlock/close operation decreases it. Then, if the reference count reaches zero, clear the lock.

IIRC, this is what ReentrantLock does.

Also, to keep one open operation from stomping on another, we should probably keep separate locks for each directory...in a map.

I think this re-proves the need for a map of ReentrantLock keyed on directory path, similar to the one FileTree uses in partyline. If you use a Map<String, ReentrantLock>, then you'll still have to synchronize the lazy initialization of a new ReentrantLock, but it's a small synchronization point. 

If you used a map of ReentrantLock, I think you could also get rid of synchronizing on Transfer and just use the ReentrantLock (probably).

NOTE: If Transfer itself has synchronized methods, you might have to implement something like:

```
    <T> T CacheProvider.synchronizeAnd( transfer, ()->{ /*some logic*/});
```

Then, in FastLocalCacheProvider:

```
    public <T> T synchronizeAnd( Transfer transfer, Supplier<T> operation){
        // lookup and lock the ReentrantLock
        return operation.get();
        // finally: release lock
    }
```

In PartylineCacheProvider we might implement as:

```
    public <T> T synchronizeAnd( Transfer transfer, Supplier<T> operation){
        // NOTE: This might need to be exposed in JoinableFileManager
        return filemanager.lockAnd( transfer.getDetachedFile(), operation);
    }
```

And in FileCacheProvider, it might be simpler:

```
    public <T> T synchronizeAnd( Transfer transfer, Supplier<T> operation){
        synchronized( transfer ){
            return operation.get();
        }
    }
```

I'm not sure how this stuff would affect the cluster, though. The lock ownership is still maintained in the cache, which is clustered. The ReentrantLock would cause local threads to wait, but remote threads should wait because the ownership is still registered in the cache. What am I missing?